File: rust/p003/src/main.rs
fn solve(mut x: u64) -> u64 {
    let mut a: u64 = 0;
    while x % 2 == 0 {
        a = 2;
        x /= 2;
    }
    for z in 3..=((x as f64).sqrt() as u64) {
        if x % z == 0 {
            a = z;
            x /= z;
        }
    }
    return if x == 1 { a } else { x };
}

fn main() {
    println!("{}", solve(600851475143));
}

File: haskell/p003/Main.hs
module Main where

solve :: Integral a => a -> a
solve a = largestPrimeFactor 1 a
  where
    largestPrimeFactor :: Integral a => a -> a -> a
    largestPrimeFactor a x
      | even x = largestPrimeFactor 2 (div x 2)
      | a <= floor (sqrt $ fromIntegral x) = largestPrimeFactor (a + 2) (if mod x a == 0 then div x a else x)
      | x == 1 = a
      | otherwise = x

main :: IO ()
main = print $ solve 600851475143
